### PR TITLE
schemas: identifier scheme refactor

### PR DESCRIPTION
**Pull request by ppanero**
closes https://github.com/inveniosoftware/marshmallow-utils/issues/37

This implementation changes the behavious of the previous and has much more strict validation. Why?
- If you are using this field to validate, it should actually validate (i.e. removes `allow_all` and similar implementation since for that there is no need to use this schema)
- Restricts the use cases, and forces the user to specify which schemes are allowed. Again, using this field means *validating* and therefore there must be some validation "rules" (schemes).

### CODE REVIEW

File: tests/schemas/test_identifier_schema.py [hunk 1]
```diff
@@ -14,231 +14,175 @@
...
+        schema.load(valid_doi)
 
-    schema = IdentifierSchema(allowed_schemes=["isbn", "orcid"])
-    data = schema.load(invalid_scheme_identifier)
 
-    with_scheme = {"scheme": "orcid"}
-    with_scheme.update(invalid_scheme_identifier)
-    assert with_scheme == data == schema.dump(data)
+def test_given_and_allowed_empty_scheme_valid_value():  # 3 and 7
+    # NOTE: does not allow the instantiation with emtpy allowed schemes
```
Comment (fenekku) [inline tests/schemas/test_identifier_schema.py:120]: :100:

File: tests/schemas/test_identifier_schema.py [hunk 2]
```diff
@@ -14,231 +14,175 @@
...
+    schema = IdentifierSchema(allowed_schemes=["isni"])
+    valid_isni = {
+        "identifier": "0000-0001-6759-6273"
     }
+    data = schema.load(valid_isni)
 
-    with pytest.raises(ValidationError):
-        CustomExtraNotRequiredSchema().load(invalid_custom)
+    valid_isni["scheme"] = "isni"
+    assert data == valid_isni
```
Comment (fenekku) [inline tests/schemas/test_identifier_schema.py:188]: If we want to cover last branch of `IdentifierSchema`'s `__init__`, we could do:

```python
def test_invalid_allowed_schemes():
    with pytest.raises(ValidationError):
        IdentifierSchema(allowed_schemes=[1, 2, 3])
```

**Review by fenekku — APPROVED** [review]
Many minor things for clarity, but overall IdentifierSchema will be much clearer to use thanks to this refactor.

File: tests/schemas/test_identifier_schema.py [hunk 3]
```diff
@@ -14,231 +14,175 @@
...
-    extra = Str()
 
-
-def test_custom_inheritance_extra_with_identifier_should_pass():
-    valid_custom = {
-        "extra": "random",
-        "identifier": "0000-0001-6759-6273",
-        "scheme": "orcid",
+def test_detected_and_allowed_scheme_respect_detection_orther():  # 8
+    # NOTE: pmid is PubMed ID
```
[the remark this answers is not in the document]
Comment (ppanero) [reply tests/schemas/test_identifier_schema.py:169]: nope, thanks

File: tests/schemas/test_identifier_schema.py (hunk 2 above)
Comment (ppanero) [reply tests/schemas/test_identifier_schema.py:188]: True, thanks! added at first (along with other more related tests)

File: marshmallow_utils/schemas/identifier.py [hunk 4]
```diff
@@ -95,26 +104,16 @@ def validate_identifier(self, data, **kwargs):
 
         if identifier and not scheme:
             raise ValidationError(
-                f"Missing scheme value for identifier {identifier}."
+                f"Missing or invalid scheme for identifier {identifier}."
             )
 
         if identifier:
```
[the remark this answers is not in the document]
Comment (ppanero) [reply marshmallow_utils/schemas/identifier.py:110]: Added a test case for it (end of the file)

**Pull request merged**